core/src/main/java/jenkins/security/apitoken/ApiTokenStore.java

```diff
@@ -176,18 +176,50 @@ private void addLegacyToken(@Nonnull Secret legacyToken, boolean migrationFromEx
         // 16x8=128bit worth of randomness, using brute-force you need on average 2^127 tries (~10^37)
         byte[] random = new byte[16];
         RANDOM.nextBytes(random);
+        // 32-char in hex
         String secretValue = Util.toHexString(random);
         String tokenTheUserWillUse = HASH_VERSION + secretValue;
         assert tokenTheUserWillUse.length() == 2 + 32;
         
-        String secretValueHashed = this.plainSecretToHashInHex(secretValue);
+        HashedToken token = prepareAndStoreToken(name, secretValue);
+        
+        return new TokenUuidAndPlainValue(token.uuid, tokenTheUserWillUse);
+    }
+    
+    /**
+     * Be careful with this method. Depending on how the tokenPlainValue was stored/sent to this method, 
+     * it could be a good idea to generate a new token randomly and revoke this one.
+     */
+    public synchronized @Nonnull String addFixedNewToken(@Nonnull String name, @Nonnull String tokenPlainValue) {
+        if (tokenPlainValue.length() != 2 + 32) {
+            LOGGER.log(Level.INFO, "addFixedNewToken, length received: {0}" + tokenPlainValue.length());
+            throw new IllegalArgumentException("The token must consist of 2 characters for the version and 32 hex-characters for the secret");
+        }
+        
+        String hashVersion = tokenPlainValue.substring(0, 2);
+        if (!HASH_VERSION.equals(hashVersion)) {
+            throw new IllegalArgumentException("The given version is not recognized: " + hashVersion);
+        }
+        
+        String tokenPlainHexValue = tokenPlainValue.substring(2);
+        tokenPlainHexValue = tokenPlainHexValue.toLowerCase();
+        if (!tokenPlainHexValue.matches("[a-f0-9]{32}")) {
```

Comment:
This should be extracted into a static Pattern field for caching the compiled regex.

Reply:
Not sure the situation is improved in terms of readability, compared to simple String constant. The caching part, esp. for such simple regex is a matter of local optimization I think.

Reply:
Generally, it's considered good practice to cache compiled patterns. I don't see why that wouldn't apply here.

Reply:
@jeffret-b FYI it was corrected with `CHECK_32_HEX_CHAR`